torch/ao/quantization/experimental/observer.py

```diff
@@ -4,44 +4,135 @@
+        self.gamma = 0.0
 
-    def _calculate_qparams(self) -> Tuple[torch.Tensor, torch.Tensor]:
-        pass
+    r""" Calculates nonuniform quantization parameters given min and max value tensors.
+    Args:
+        min_val: Minimum values per channel
+        max_val: Maximum values per channel
+    Returns:
+        gamma: gamma quantization parameter, defined to ensure that alpha is the maximum of the range
+        quantization_levels: non-uniform quantization levels, calculated according to APoT paper:
+        (https://arxiv.org/pdf/1909.13144.pdf)
+        level_indices: int representation of quantization_levels
+    """
+    def _calculate_qparams(
+        self,
+        min_val: torch.tensor,
+        max_val: torch.tensor,
+            signed: bool) -> Tuple[float, torch.tensor, torch.tensor]:
+        # compute alpha
+        self.alpha = max_val
 
-class APoTObserver(NonUniformQuantizationObserverBase):
-    alpha = 0
-    gamma = 0
-    level_indices = torch.Tensor()
+        # compute n and store as member variable
+        if self.k:
+            self.n = self.b // self.k
+
+        if self.n == 0:
+            return (0.0, torch.tensor([]), torch.tensor([]))
+
+        # store a tensor of subtensors (all levels)
+        p_all = []
+
+        # create levels
+        for i in range(0, self.n):
+            p_curr = torch.tensor([0])
+
+            for j in range(0, 2 ** (self.k - 1) + 1):
+                curr_ele = 2 ** (- (i + j * self.n))
+                p_append = torch.tensor([curr_ele])
+                p_curr = torch.cat((p_curr, p_append))
+                # introduce signed numbers
+                if signed:
+                    p_curr = torch.cat((p_curr, torch.tensor([-curr_ele])))
+
+            if signed:
+                # sort tensor in reverse order before adding to list if signed
+                sorted, indices = torch.sort(p_curr, descending=True)
+                p_all.append(sorted)
+            else:
+                p_all.append(p_curr)
+
+        # gamma calculation:
+        # loop through all tensors, add element at index 1 for each tensor
+        p_sum = 0
+        for tens in p_all:
+            p_sum += tens[1]
 
+        # assign gamma
+        self.gamma = self.alpha / p_sum
+
+        # calculate cartesian product
+        cartesian_product = list(itertools.product(*p_all))
+
+        quantization_levels = []
+
+        # calculate sum of each row
+        for row in cartesian_product:
+            sum = 0
+            for ele in row:
+                sum += ele
+            quantization_levels.append(sum)
+
+        quantization_levels = [self.gamma * ele for ele in quantization_levels]
+        quantization_levels = torch.tensor(quantization_levels)
+        quantization_levels, level_indices = quantization_levels.sort()
+
+        return (self.gamma, quantization_levels, level_indices)
+
+class APoTObserver(NonUniformQuantizationObserverBase):
     def __init__(
```

Comment:
needs a description of the API vars meanings

Reply:
alternatively more descriptive variable names

Reply:
@asl3  please link the paper and describe what the variables mean. I only know what gamma means because I read the paper

Reply:
oh never mind, I see you did that above